src/uct/tcp/tcp_ep.c

```diff
@@ -861,17 +916,54 @@ ucs_status_t uct_tcp_ep_am_short(uct_ep_h uct_ep, uint8_t am_id, uint64_t header
 
     ucs_assertv(hdr != NULL, "ep=%p", ep);
 
-    *((uint64_t*)(hdr + 1)) = header;
-    memcpy((uint8_t*)(hdr + 1) + sizeof(header), payload, length);
     /* Save the length of the payload, because hdr (ep::buf)
      * can be released inside `uct_tcp_ep_am_send` call */
     hdr->length = payload_length = length + sizeof(header);
 
-    uct_tcp_ep_am_send(iface, ep, hdr);
+    if (length <= iface->config.min_am_shortv) {
+        uct_tcp_ep_am_short_fill_data(hdr, header, payload, length);
 
-    UCT_TL_EP_STAT_OP(&ep->super, AM, SHORT, payload_length);
+        uct_tcp_ep_am_send(iface, ep, hdr);
+        UCT_TL_EP_STAT_OP(&ep->super, AM, SHORT, payload_length);
+    } else {
+        iov[0].iov_base = hdr;
+        iov[0].iov_len  = sizeof(*hdr);
+
+        iov[1].iov_base = &header;
+        iov[1].iov_len  = sizeof(header);
+
+        iov[2].iov_base = (void*)payload;
+        iov[2].iov_len  = length;
+
+        status = uct_tcp_ep_am_sendv(iface, ep, 1, hdr,
+                                     iface->config.tx_seg_size, &header,
+                                     iov, UCT_TCP_EP_AM_SHORTV_IOV_COUNT);
```

Comment:
here is potential issue: in case if macro UCT_TCP_EP_AM_SHORTV_IOV_COUNT will be changed for some reason this code will be incorrect. what about use counter variable how many iovs are filled? to be sure that iov_count is always correct

@brminich what do you think?

Reply:
let's wait @brminich answer, but I think there is no good way here. If we will use counter, we could possibly miss incrementing it or something like that.
Currently, if `UCT_TCP_EP_AM_SHORTV_IOV_COUNT` value would be changed for some reason, this will affect `iov` array declared above

Reply:
imo, either way is ok (macro or constant/counter)